Html sanitizer adds rel="nofollow" even if it pre-exists

Hi,

Consider the input as 
String input = ```html <a href='www.google.com' rel='nofollow'></a>```;

In my policy factory i have requireRelNofollowOnLinks() ON.

Expected output : ```html <a href='www.google.com' rel='nofollow'></a> ```;

Actual output : ```html <a href="www.google.com" rel="nofollow" rel="nofollow"></a>```

Sometimes it adds multiple rel="nofllow" which will result in more than 2 rel="follow" in single tag.

Thanks

Here is a failing test case:

```java
package org.riverporpoise;

import static org.junit.Assert.assertEquals;

import org.junit.Test;
import org.owasp.html.HtmlPolicyBuilder;
import org.owasp.html.PolicyFactory;

public class AppTest 
{
    @Test
    public void issue145()
    {
		final PolicyFactory policy = new HtmlPolicyBuilder().allowElements("a")
				.allowAttributes("href", "target", "rel").onElements("a")
				.allowStandardUrlProtocols()
				.requireRelNofollowOnLinks()
				.toFactory();
		final String htmlContent = "<a href=\"http://google.com/\" rel=\"nofollow noopener noreferrer\" target=\"_blank\">hello</a>";
		final String sanitized = policy.sanitize(htmlContent);
		assertEquals(htmlContent, sanitized);
    }
}
```

This outputs:
```
[INFO] Running org.riverporpoise.AppTest
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.153 s <<< FAILURE! - in org.riverporpoise.AppTest
[ERROR] issue145(org.riverporpoise.AppTest)  Time elapsed: 0.107 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<... noopener noreferrer[]" target="_blank">he...> but was:<... noopener noreferrer[ nofollow noopener noreferrer]" target="_blank">he...>
	at org.riverporpoise.AppTest.issue145(AppTest.java:21)
```

All three values are duplicated.

Note that I have prescribed specific sequencing of rel values in my test, for brevity's sake.

> Here is a failing test case:
> 
> ```java
> package org.riverporpoise;
> 
> import static org.junit.Assert.assertEquals;
> 
> import org.junit.Test;
> import org.owasp.html.HtmlPolicyBuilder;
> import org.owasp.html.PolicyFactory;
> 
> public class AppTest 
> {
>     @Test
>     public void issue145()
>     {
> 		final PolicyFactory policy = new HtmlPolicyBuilder().allowElements("a")
> 				.allowAttributes("href", "target", "rel").onElements("a")
> 				.allowStandardUrlProtocols()
> 				.requireRelNofollowOnLinks()
> 				.toFactory();
> 		final String htmlContent = "<a href=\"http://google.com/\" rel=\"nofollow noopener noreferrer\" target=\"_blank\">hello</a>";
> 		final String sanitized = policy.sanitize(htmlContent);
> 		assertEquals(htmlContent, sanitized);
>     }
> }
> ```
> 
> This outputs:
> 
> ```
> [INFO] Running org.riverporpoise.AppTest
> [ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.153 s <<< FAILURE! - in org.riverporpoise.AppTest
> [ERROR] issue145(org.riverporpoise.AppTest)  Time elapsed: 0.107 s  <<< FAILURE!
> org.junit.ComparisonFailure: expected:<... noopener noreferrer[]" target="_blank">he...> but was:<... noopener noreferrer[ nofollow noopener noreferrer]" target="_blank">he...>
> 	at org.riverporpoise.AppTest.issue145(AppTest.java:21)
> ```
> 
> All three values are duplicated.
> 
> Note that I have prescribed specific sequencing of rel values in my test, for brevity's sake.

@dhorlick dhorlick Is your issue resolved? Please help me on getting it resolved

--- a/src/main/java/org/owasp/html/HtmlPolicyBuilder.java
+++ b/src/main/java/org/owasp/html/HtmlPolicyBuilder.java
@@ -1049,7 +1049,7 @@ public class HtmlPolicyBuilder {
                     final String lowerCaseRel = Strings.toLowerCase(rel);
                     if ((skip.isEmpty() || !skip.contains(lowerCaseRel)) && !present.contains(lowerCaseRel)) {
                       present.add(lowerCaseRel);
-                      sb.append(rel).append(' ');
+                      sb.append(lowerCaseRel).append(' ');
                     }
                   }
                   left = i + 1;

--- a/src/main/java/org/owasp/html/HtmlPolicyBuilder.java
+++ b/src/main/java/org/owasp/html/HtmlPolicyBuilder.java
@@ -1045,11 +1045,10 @@ public class HtmlPolicyBuilder {
               for (int i = 0; i <= n; ++i) {
                 if (i == n || Strings.isHtmlSpace(rels.charAt(i))) {
                   if (left < i) {
-                    if (skip.isEmpty()
-                        || !skip.contains(
-                            Strings.toLowerCase(rels.substring(left, i)))) {
-                      String rel = rels.substring(left, i);
-                      present.add(rel);
+                    final String rel = rels.substring(left, i);
+                    final String lowerCaseRel = Strings.toLowerCase(rel);
+                    if ((skip.isEmpty() || !skip.contains(lowerCaseRel)) && !present.contains(lowerCaseRel)) {
+                      present.add(lowerCaseRel);
                       sb.append(rel).append(' ');
                     }
                   }

--- a/src/test/java/org/owasp/html/HtmlPolicyBuilderTest.java
+++ b/src/test/java/org/owasp/html/HtmlPolicyBuilderTest.java
@@ -875,6 +875,31 @@ public class HtmlPolicyBuilderTest extends TestCase {
   }
 
   @Test
+  public static final void testRelLinksWithDuplicateRels() {
+    PolicyFactory pf = new HtmlPolicyBuilder()
+        .allowElements("a")
+        .allowAttributes("href").onElements("a")
+        .allowAttributes("rel").onElements("a")
+        .allowAttributes("target").onElements("a")
+        .allowStandardUrlProtocols()
+        .toFactory();
+    assertEquals("<a target=\"_blank\" rel=\"noopener noreferrer\" href=\"https://google.com\">test</a>", pf.sanitize("<a target=\"_blank\" rel=\"noopener noreferrer noreferrer\" href=\"https://google.com\">test</a>"));
+  }
+
+  @Test
+  public static final void testRelLinksWithDuplicateRelsRequired() {
+    PolicyFactory pf = new HtmlPolicyBuilder()
+        .allowElements("a")
+        .allowAttributes("href").onElements("a")
+        .allowAttributes("rel").onElements("a")
+        .allowAttributes("target").onElements("a")
+        .allowStandardUrlProtocols()
+        .requireRelsOnLinks("noreferrer")
+        .toFactory();
+    assertEquals("<a target=\"_blank\" rel=\"noopener noreferrer\" href=\"https://google.com\">test</a>", pf.sanitize("<a target=\"_blank\" rel=\"noopener noreferrer noreferrer\" href=\"https://google.com\">test</a>"));
+  }
+
+  @Test
   public static final void testFailFastOnSpaceSeparatedStrings() {
     boolean failed;
     try {

